packages/editor/src/editor.ts

```diff
@@ -0,0 +1,236 @@
+      const hasMongodbExtension = !!extensions
+        .find((name) => name.includes('mongodb.mongodb-vscode'));
+
+      this.messageBus.emit('mongosh-editor:read-vscode-extensions-done', {
+        vscodeDir: this._vscodeDir,
+        hasMongodbExtension
+      });
+
+      return hasMongodbExtension ? 'mongodb' : 'js';
+    } catch (error) {
+      this.messageBus.emit('mongosh-editor:read-vscode-extensions-failed', {
+        action: 'mongosh-editor:read-vscode-extensions-failed',
+        error: error.message
+      });
+
+      return 'js';
+    }
+  }
+
+  async _getEditor(): Promise<string|null> {
+    // Check for an external editor in the mongosh configuration.
+    let editor: string | null = await this._instanceState.shellApi.config.get('editor');
+
+    // Check for an external editor in the environment variable.
+    if (!editor && process.env.EDITOR) {
+      editor = process.env.EDITOR;
+    }
+
+    return editor;
+  }
+
+  async _createTempFile(): Promise<void> {
+    const ext = await this._getExtension(this._editorSession.cmd);
+    const content = await this._getEditorContent();
+
+    this._editorSession.tmpDoc = path.join(this._tmpDir, `edit-${uuidv4()}.${ext}`);
+
+    // Create a temp file to store content that is being edited.
+    await fse.ensureFile(this._editorSession.tmpDoc);
+    await fse.writeFile(this._editorSession.tmpDoc, beautify(content));
+  }
+
+  async _readTempFile(tmpDoc: string): Promise<string> {
+    const content = await fse.readFile(tmpDoc, 'utf8');
+    await fse.unlink(tmpDoc);
+    return this._makeMultilineJSIntoSingleLine(content);
+  }
+
+  _isVscodeApp(cmd: string): boolean {
+    const regex = /^((.*[\/\\])*)?[cC]ode(.exe)?$/;
+
+    try {
+      return !!regex.exec(cmd);
+    } catch (error) {
+      this.messageBus.emit('mongosh-editor:is-vscode-app-regex-failed', { regex, cmd });
+      return false;
+    }
+  }
+
+  _isStatement(code: string): boolean {
+    const regex = /\b[^()]+\((.*)\)$/;
+
+    try {
+      return !!regex.exec(code);
+    } catch (error) {
+      this.messageBus.emit('mongosh-editor:is-statement-regex-failed', { regex, code });
+      return false;
+    }
+  }
+
+  async _getEditorContent(): Promise<string> {
+    if (!this._editorSession.code) {
+      return this._editorSession.content;
+    }
+
+    const evalResult = await this._loadExternalCode(this._editorSession.code, '@(shell eval)');
+    return this._isStatement(this._editorSession.code) ? this._editorSession.code : evalResult.toString();
+  }
+
+  async runEditCommand([ code, ...args ]: string[]): Promise<void> {
+    await this.print('Opening an editor...');
+    const editor: string|null = await this._getEditor();
+
+    // If none of the above configurations are found return an error.
+    if (!editor) {
+      throw new Error(`Command failed: ${[code, ...args].join(' ')} with an error: please define an external editor`);
+    }
+
+    const [ cmd, ...cmdArgs ] = editor.split(' ');
```

Comment:
I think this is going to be problematic on Windows, where the path to the editor can contain spaces as well.

(i.e.: There is effectively no good way for us to tell where in `C:\Program Files\Editor\editor.exe --flag` the path to the binary itself ends and where other arguments begin, and we’ll need the system shell to handle that.)